src/util/server.ml

```diff
@@ -0,0 +1,155 @@
+
+let serve serv =
+  let chan = IO.to_input_channel stdin in
+  let stream = Yojson.Safe.linestream_from_channel chan in
+  while not (Stream.is_empty stream) do
+    let line = Stream.next stream in
+    match line with
+    | `Json json -> (
+        try
+          let message = Message.either_of_yojson json in
+          match message.id with
+          | Some id -> handle_request serv message id
+          | _ -> () (* We just ignore notifications for now. *)
+        with exn -> prerr_endline (Printexc.to_string exn))
+    | `Exn exn -> prerr_endline (Printexc.to_string exn)
+  done
+
+let make file do_analyze : t = { file; do_analyze }
+
+let start file do_analyze =
+  GobConfig.set_bool "incremental.save" true;
+  serve (make file do_analyze)
+
+let analyze ?(reset=false) { file; do_analyze } =
+  if reset then (
+    Serialize.solver_data := None;
+    Messages.Table.(MH.clear messages_table);
+    Messages.Table.messages_list := []);
+  let increment_data, fresh = match !Serialize.solver_data with
+    | Some solver_data ->
+      let changes = CompareCIL.compareCilFiles file file in
+      let old_data = Some { Analyses.cil_file = file; solver_data } in
+      { Analyses.changes; old_data; new_file = file }, false
+    | _ -> Analyses.empty_increment_data file, true
+  in
+  GobConfig.set_bool "incremental.load" (not fresh);
+  do_analyze increment_data file;
+  GobConfig.set_bool "incremental.load" true
+
+let () =
+  let register = Registry.register registry in
+
+  register (module struct
+    let name = "analyze"
+    type args = { reset: bool [@default false] } [@@deriving of_yojson]
+    type result = unit [@@deriving to_yojson]
+    let process { reset } serve = analyze serve ~reset
+  end);
+
+  register (module struct
+    let name = "config"
+    type args = string * Yojson.Safe.t [@@deriving of_yojson]
+    type result = unit [@@deriving to_yojson]
+    let process (conf, json) _ = GobConfig.set_auto conf (Yojson.Safe.to_string json)
+  end);
+
+  register (module struct
+    let name = "merge_config"
+    type args = Yojson.Safe.t [@@deriving of_yojson]
+    type result = unit [@@deriving to_yojson]
+    let process json _ = GobConfig.merge json
+  end);
+
+  register (module struct
+    let name = "messages"
+    type args = unit [@@deriving of_yojson]
+    type result = Messages.Message.t list [@@deriving to_yojson]
+    let process () _ = !Messages.Table.messages_list
+  end);
+
+  register (module struct
+    let name = "exp_eval"
+    type args = ExpressionEvaluation.query [@@deriving of_yojson]
+    type result =
+      ((string * CilType.Location.t * string * int) * bool option) list [@@deriving to_yojson]
+    let process query serv =
+      GobConfig.set_auto "trans.activated[+]" "'expeval'";
+      ExpressionEvaluation.gv_query := Some query;
+      analyze serv;
+      GobConfig.set_auto "trans.activated[-]" "'expeval'";
+      !ExpressionEvaluation.gv_results
+  end);
+
```

Comment:
Do you have a full list of the possible commands we discussed at the server mode meeting a while ago? Even if they're not all implemented yet, maybe it would be a good idea to leave TODOs here with the ones that are still missing. Would make it easier to know right from the code instead of having to find some external document or list.

Reply:
@keremc I posted a list of these on the #interactive slack channel.

Reply:
Thanks, I added some TODOs in the code.